# Project 3 Level 1
# By Devansh Panchal

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

##### importing necessary libraries

In [7]:
df = pd.read_csv("C:/Users/91812/Desktop/Internship/Oasis Infobyte/Datasets/AB_NYC_2019.csv")

#### Importing the data & getting Insights about it

In [13]:
df.head()

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
0,2539,Clean & quiet apt home by the park,2787,John,Brooklyn,Kensington,40.64749,-73.97237,Private room,149,1,9,2018-10-19,0.21,6,365
1,2595,Skylit Midtown Castle,2845,Jennifer,Manhattan,Midtown,40.75362,-73.98377,Entire home/apt,225,1,45,2019-05-21,0.38,2,355
2,3647,THE VILLAGE OF HARLEM....NEW YORK !,4632,Elisabeth,Manhattan,Harlem,40.80902,-73.94190,Private room,150,3,0,NaN,NaN,1,365
3,3831,Cozy Entire Floor of Brownstone,4869,LisaRoxanne,Brooklyn,Clinton Hill,40.68514,-73.95976,Entire home/apt,89,1,270,2019-07-05,4.64,1,194
4,5022,Entire Apt: Spacious Studio/Loft by central park,7192,Laura,Manhattan,East Harlem,40.79851,-73.94399,Entire home/apt,80,10,9,2018-11-19,0.10,1,0


In [12]:
print(df.shape)

(48895, 16)


#### The dataset containes 48,895 Records with their 16 Features

In [17]:
print(df.dtypes)

id                                  int64
name                                  str
host_id                             int64
host_name                             str
neighbourhood_group                   str
neighbourhood                         str
latitude                          float64
longitude                         float64
room_type                             str
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                           str
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object


In [18]:
print(df.isna().sum())

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64


#### Here we can find the null values which are present in the attributes (name: 16, host_name: 21, last_review & reviews_per_month: 10052)

In [21]:
df.describe()

,id,host_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365
count,4.889500e+04,4.889500e+04,48895.000000,48895.000000,48895.000000,48895.000000,48895.000000,38843.000000,48895.000000,48895.000000
mean,1.901714e+07,6.762001e+07,40.728949,-73.952170,152.720687,7.029962,23.274466,1.373221,7.143982,112.781327
std,1.098311e+07,7.861097e+07,0.054530,0.046157,240.154170,20.510550,44.550582,1.680442,32.952519,131.622289
min,2.539000e+03,2.438000e+03,40.499790,-74.244420,0.000000,1.000000,0.000000,0.010000,1.000000,0.000000
25%,9.471945e+06,7.822033e+06,40.690100,-73.983070,69.000000,1.000000,1.000000,0.190000,1.000000,0.000000
50%,1.967728e+07,3.079382e+07,40.723070,-73.955680,106.000000,3.000000,5.000000,0.720000,1.000000,45.000000
75%,2.915218e+07,1.074344e+08,40.763115,-73.936275,175.000000,5.000000,24.000000,2.020000,2.000000,227.000000
max,3.648724e+07,2.743213e+08,40.913060,-73.712990,10000.000000,1250.000000,629.000000,58.500000,327.000000,365.000000


In [27]:
print(df.duplicated().sum())

0


#### There aren't any duplicates in the given dataset

In [28]:
df = df.dropna(subset=['name', 'host_name'])

#### Removing the records where the values of 'name' or 'host_name' are null

In [29]:
df.isna().sum()

id                                    0
name                                  0
host_id                               0
host_name                             0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10037
reviews_per_month                 10037
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [34]:
print(df[['number_of_reviews','last_review','reviews_per_month']])


       number_of_reviews last_review  reviews_per_month
0                      9  2018-10-19               0.21
1                     45  2019-05-21               0.38
2                      0         NaN                NaN
3                    270  2019-07-05               4.64
4                      9  2018-11-19               0.10
...                  ...         ...                ...
48890                  0         NaN                NaN
48891                  0         NaN                NaN
48892                  0         NaN                NaN
48893                  0         NaN                NaN
48894                  0         NaN                NaN

[48858 rows x 3 columns]


#### Here we can see that the fields where the number_of_reviews are '0', have the respected attributes last_review & reviews_per_month as NULL

In [35]:
# Rename 'last_review' to 'last_review_date'
df.rename(columns={'last_review': 'last_review_date'}, inplace=True)

#### changing the attribute name: last_review to last_review_date for better understanding 

In [36]:
df['last_review_date'] = pd.to_datetime(df['last_review_date'], errors='coerce')

#### converting the attribute last_review_date to pandas date-time type

In [43]:
print(df['last_review_date'].dtype)

datetime64[us]


In [44]:
df['reviews_per_month'] = df['reviews_per_month'].fillna(0)

#### imputing the null values of reviews_per_month with 0 as there aren't any reviews given

In [45]:
df.isnull().sum()

id                                    0
name                                  0
host_id                               0
host_name                             0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review_date                  10037
reviews_per_month                     0
calculated_host_listings_count        0
availability_365                      0
dtype: int64

#### Here we can see 10037 null values in 'last_review_date' attribute are there aren't any reviews given.

#### We are adding a feature 'has_reviewed' which will be a flag indicating that the properity has ever been reviewed or not

In [47]:
df['has_reviewed'] = (df['number_of_reviews'] > 0).astype(int)

In [52]:
print(df[['number_of_reviews', 'last_review_date', 'reviews_per_month', 'has_reviewed']].head(10))

   number_of_reviews last_review_date  reviews_per_month  has_reviewed
0                  9       2018-10-19               0.21             1
1                 45       2019-05-21               0.38             1
2                  0              NaT               0.00             0
3                270       2019-07-05               4.64             1
4                  9       2018-11-19               0.10             1
5                 74       2019-06-22               0.59             1
6                 49       2017-10-05               0.40             1
7                430       2019-06-24               3.47             1
8                118       2017-07-21               0.99             1
9                160       2019-06-09               1.33             1


In [ ]:
print(df.describe())

                 id       host_id      latitude     longitude         price  \
count  4.885800e+04  4.885800e+04  48858.000000  48858.000000  48858.000000   
mean   1.902335e+07  6.763169e+07     40.728941    -73.952170    152.740309   
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000   
25%    9.475980e+06  7.818669e+06     40.690090    -73.983070     69.000000   
50%    1.969114e+07  3.079133e+07     40.723070    -73.955680    106.000000   
75%    2.915765e+07  1.074344e+08     40.763107    -73.936280    175.000000   
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000   
std    1.098289e+07  7.862389e+07      0.054528      0.046159    240.232386   

       minimum_nights  number_of_reviews            last_review_date  \
count    48858.000000       48858.000000                       38821   
mean         7.012444          23.273098  2018-10-04 07:23:47.072976   
min          1.000000           0.000000         2011-03-28 00:00:00   


#### As we can see that the 'price' attribute is highly Right skewed so we need to do log transform first

#### For Scaling purpose we use the RoustScaler because it utilizes median and IQR, as the StandardScaler uses mean to scale, which in this case would be high due to the max value

In [60]:
df['price'] = np.log1p(df['price'])

In [61]:
from sklearn.preprocessing import RobustScaler

scaler = RobustScaler()

cols = [
    'latitude',
    'longitude',
    'price',
    'minimum_nights',
    'number_of_reviews',
    'reviews_per_month',
    'calculated_host_listings_count',
    'availability_365'
]

df[cols] = scaler.fit_transform(df[cols])

In [63]:
df.to_csv("AB_NYC_2019_cleaned_scaled.csv", index=False)

## Things we have done so far

### Checked for duplicates<br>Removed records having the null values in name or host_name<br>Changed the name of last_review to last_review_date and it's data_type to pandas DateTime type<br>Imputed the null values of the column: reviews_per_month<br>Added a flag feature 'has_reviewed' that tells us that the property has been reviewed or not<br>Managed the outliers and Scaled the data<br>Created updated Dataset